File: python/medium/find_binary_node.py
data = {
    "value": 10,
    "left": {
        "value": 6,
        "left": {
            "value": 4,
            "left": False,
            "right": False,
        },
        "right": {
            "value": 8,
            "left": False,
            "right": False,
        },
    },
    "right": {
        "value": 18,
        "left": {
            "value": 15,
            "left": False,
            "right": False,
        },
        "right": {
            "value": 21,
            "left": False,
            "right": False,
        },
    },
}
def find_binary_node(binary_tree, value):
    node = binary_tree
    while node:
        if node["value"] == value:
            return node
        elif value > node["value"]:
            node = node["right"]
        else:
            node = node["left"]
    return "NOT FOUND"
print(find_binary_node(data,21))
print(find_binary_node(data,4))
print(find_binary_node(data,9))

def insert_last_binary_node(binary_tree, new_node):
    node = binary_tree
    while node:
        last_node = node
        if new_node["value"] > node["value"]:
            node = node["right"]
        else:
            node = node["left"]
    if new_node["value"] > last_node["value"]:
        last_node["right"] = new_node
    else:
        last_node["left"] = new_node
insert_last_binary_node(data,{"value": 29, "left": False, "right": False})
print(find_binary_node(data,29))
print(data)
#   O(logN) for balanced only

File: ts/medium/findBinaryNode.ts
// `
//                                 root(5)
//                              ______|______
//                             /             \
//                         parent(4)       parent(7)
//                          /  \             /    \
//           internal node(2)  leaf(3)  leaf(6)  leaf(8)
//                     /
//                  leaf(1)

// tree sum: 36
// tree depth: 4
// tree height: 6
// `;

type BTreeNode = {
  value: number;
  left: BTreeNode | null;
  right: BTreeNode | null;
};

type BTree = BTreeNode | null;

export const bTree: BTree = {
  // root
  value: 5,

  // parent
  left: {
    value: 4,

    // Internal node
    left: {
      value: 2,
      // leaf
      left: {
        value: 1,
        left: null,
        right: null,
      },
      right: null,
    },

    // leaf
    right: {
      value: 3,
      left: null,
      right: null,
    },
  },

  // parent
  right: {
    value: 7,

    // leaf
    left: {
      value: 6,
      left: null,
      right: null,
    },

    // leaf
    right: {
      value: 8,
      left: null,
      right: null,
    },
  },
};

function findBinaryNode(binaryTree: BTree, target: number): BTreeNode | null {
  let node: BTreeNode | null = binaryTree;
  while (node) {
    if (node.value === target) {
      return node;
    } else if (target > node.value) {
      node = node.right;
    } else {
      node = node.left;
    }
  }
  return null;
}
console.log(findBinaryNode(bTree, 7));
console.log(findBinaryNode(bTree, 23));

function insertLastToBinaryNode(binaryTree: BTree, newNode: BTreeNode): void {
  let node: BTreeNode | null = binaryTree;
  let lastNode: BTreeNode | null = null;

  while (node) {
    lastNode = node;
    if (newNode.value > node.value) {
      node = node.right;
    } else if (newNode.value < node.value) {
      node = node.left;
    } else {
      console.warn("Node with the same value already exists.");
      return;
    }
  }
  
  if (lastNode) {
    if (newNode.value > lastNode.value) {
      lastNode.right = newNode;
    } else {
      lastNode.left = newNode;
    }
  }
}
insertLastToBinaryNode(bTree, { value: 10, left: null, right: null });
console.log('bTree: ', JSON.stringify(bTree));